# Using Fasta Input
<!-- WARNING -->
<!-- Do not edit this file from within the docs.nextstrain.org repository. -->
<!-- It is fetched from another repository to be included in the docs.nextstrain.org build. -->
<!-- So, if you edit it after it is fetched into docs.nextstrain.org, your changes will be lost. -->
<!-- Instead, edit this file in its own repository and commit your changes there. -->
<!-- For more details on this (temporary) implementation, see https://github.com/nextstrain/docs.nextstrain.org#fetching-of-documents-from-other-repositories -->
<!-- This file is fetched from: https://github.com/nextstrain/augur/blob/master/docs/faq/fasta_input.md -->
<!-- WARNING -->
<!-- WARNING -->
<!-- WARNING -->

If you would like to run augur on viral data, or bacterial SNP data, you probably would like to start with Fasta sequence data.

### Sequence data

Your sequence data should

* consist of homologous sequences that can be aligned unambiguously
* needs to contain sufficient diversity to allow reliable tree reconstruction
* should be of similar length. Mixing short sequences (300bp) with much longer ones (10000bp) often yields unexpected results.
